Changes `selector-no-type` to automatically resolve nested selectors when `ignore: ["descendant"]` option is enabled.

manovotny opened the pull request:
### Description

This PR addresses #1654.

These changes will automatically resolve nested selectors automatically if the `ignore: ["descendant"]` option is enabled. 

And since this is my first PR on this project, please let me know if there's anything else I may have missed or of there is any other pertinent information I should add to the PR.

### Benchmarks

#### Before

```
❯ npm run benchmark-rule -- selector-no-type "[true, {\"ignore\": [\"descendant\"]}]"

Warnings: 406
Mean: 171.1881831515151 ms
Deviation: 38.763392889220015 ms
```

#### After

```
❯ npm run benchmark-rule -- selector-no-type "[true, {\"ignore\": [\"descendant\"]}]"

Warnings: 406
Mean: 156.6367916 ms
Deviation: 25.370745159382917 ms
```

ntwb commented:
@manovotny Thanks, we'll need a changelog entry similar to [this commit](https://github.com/stylelint/stylelint/commit/c7bb78cebb3d2a14ef52669bb36104951b20c4c2#diff-4ac32a78649ca5bdd8e0ba38b7006a1e)

http://stylelint.io/developer-guide/rules/ also includes a few checklists when creating/updating rules

manovotny commented:
@ntwb : : 

Fix added to changelog.

I reviewed the [checklist](http://stylelint.io/developer-guide/rules/#checklist) and I *think* I've got everything covered. Please let me know if you think I've missed something.

ntwb commented:
@manovotny Can you add some SCSS syntax tests, similar [to these](https://github.com/stylelint/stylelint/blob/master/src/rules/selector-class-pattern/__tests__/index.js#L118-L133) please

jeddy3 commented:
@manovotny Thanks for contributing! Looking good so far.

I think we'll need to tweak the implementation a little though. In https://github.com/stylelint/stylelint/issues/1654#issuecomment-232754442 I said:

> I don't think we need to add a 2nd option this time as automatically resolving the nesting for an ignore secondary option won't be a breaking change.

And then @davidtheclark [categorised](https://github.com/stylelint/stylelint/issues/1654#issuecomment-232837182) as a bug accordingly.

As such, I think we can do away with the `resolveNestedSelectors` *option*. Instead, resolving the nesting is triggered automatically whenever the `ignore: []` secondary option is used. This change in default behaviour won't be a breaking change as we're making the rule more permissive.

manovotny commented:
@jeddy3 : : 

Oh... I'm a dope. I totally misunderstood what you said. When you said `I don't think we need to add a 2nd option`, I thought you meant we didn't need to add something alongside `compounded` and  `descendant`, which I guess would be 3rd option. 

Anyway. My bad. 

We're on the same page now. I'll get those changes and the `syntax: "scss"` changes @ntwb wanted in the tests.

Thanks!

manovotny commented:
Ok, here's were we're at...

I am now resolving nested selectors automatically if the `ignore: ["descendant"]` option is enabled. I am only doing this for `descendant` because `compounded` will work as expected regardless if it's nested or not. 

And because nesting is only possible with a preprocessor, I've changed the tests to reflect that as well with `syntax: "scss"` and `syntax: "less"`.

I have also updated the title and description of the PR (including benchmarks) and the CHANGELOG to reflect these changes too.

I consider the PR ready for review again, whenever you have time.

Thanks for the help and patience.

jeddy3 commented:
The rule code looks good to me :) We'll wait on a 2nd pair of eyes to double check that bit.

> And because nesting is only possible with a preprocessor, I've changed the tests to reflect that as well with syntax: "scss" and syntax: "less".

Thanks for making the effort, but [when we say](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#write-tests): "Ensure you use standard CSS syntax by default, and only swap parsers when testing a specific piece of non-standard syntax." we are referring to non-standard syntax like operators, `//` comments, maps and interpolation. Nesting conforms to standard CSS syntax e.g.:

```css
@media (min-width: 100px) {
  a {
    top: 10px
  }
}
```

And why there is [this proposal](https://tabatkins.github.io/specs/css-nesting/) and [corresponding PostCSS plugin](https://github.com/jonathantneal/postcss-nesting).

@ntwb's [request](https://github.com/stylelint/stylelint/pull/1659#issuecomment-232851914) was for an interpolation test e.g. `@for $n { .a { div#{$n} { } }}`.

So, just one passing `.foo { div {} }` test in the main `testRule` and one passing interpolation test in the `scss` block will suffice. You can then remove your other tests from the syntax blocks.

Don't worry, everyone's first PR goes back and forth because there's a lot of subtleties that we've evolved over the last year, especially dealing with the complexities of non-standard syntax. We've tried to document as much of it as we can, but even the documentation is subtle sometimes :)

manovotny commented:
@jeddy3 : : 

No, this is very good. I appreciate your patience. 

I am trying to be helpful and add value to the project and trying not to suck up all of your valuable time. I appreciate the guidance and willingness to kindly nudge me in the right directions.

Yes, there are lots of subtleties, but what project doesn't have them? 😄 Part of the fun / challenge (in a good way) is figuring all that out in open source projects. 

I'll take another stab at the tests and let you know when I have them in a place where think they're ready for review again.

davidtheclark commented on CHANGELOG.md:
```diff
@@ -1,6 +1,7 @@
 # Head
 
 - Fixed: bug causing rules in extended configs to be merged with, rather than replaced by, the extending config.
+- Fixed: bug in `selector-no-type` when using `ignore: ["descendant"]`, which now will automatically resolve nested selectors before checking the rule.
```
Let's rephrase/simplify to

> Fixed: `selector-no-type` with secondary option `ignore: ["descendant"]` will now resolve nested selectors.

manovotny replied:
Great suggestion! I'll change it.

davidtheclark commented on src/rules/selector-no-type/__tests__/index.js:
```diff
@@ -158,6 +158,24 @@ testRule(rule, {
+  reject: [{
+    code: ".foo { div {} }",
+    description: "nested descendant",
+    message: messages.rejected,
```
Can we add the warning position to these rejected cases to ensure that it's correct?

manovotny replied:
I can do that.

davidtheclark commented:
Thanks @manovotny! I just added a couple of comments.

manovotny commented:
Ok, @jeddy3, @ntwb, and @davidtheclark. Latest changes are ready for review.

For the tests, I believe I covered all the permutations.

* `selector-no-type` enabled, rejected
* `selector-no-type` enabled and `ignore: ["descendant"]` enabled, accepted
* `selector-no-type` enabled, scss interpolation, rejected
* `selector-no-type` enabled and `ignore: ["descendant"]` enabled, scss interpolation, accepted
* `selector-no-type` enabled, less interpolation, rejected
* `selector-no-type` enabled and `ignore: ["descendant"]` enabled, less interpolation, accepted

Please let me know if there is anything else. And please let me know if I am being dense and simply not understating what you want from a testing perspective.

jeddy3 commented:
@manovotny Thanks!

Looks thorough to me.

ntwb commented:
Thanks @manovotny 

LGTM

davidtheclark commented:
Nicely done.

manovotny commented:
Awesome! 

Thanks for all the help, @jeddy3, @ntwb, and @davidtheclark!

manovotny commented:
Any ETA on when the next release will be cut?

No pressure. Just curious.

Thanks!

jeddy3 commented:
We were [waiting on](https://github.com/stylelint/stylelint/issues/1662) one last fix: https://github.com/stylelint/stylelint/pull/1639

It's in now and so we'll have a release out in the next few minutes :)

manovotny commented:
@jeddy3 : : 

Fantastic! Thanks!
